api/v1alpha1/etcdcluster_types.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// EtcdClusterSpec defines the desired state of EtcdCluster.
+type EtcdClusterSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Size is the expected size of the etcd cluster.
+	Size int `json:"size"`
+	// Version is the expected version of the etcd container image.
+	Version string `json:"version"`
```

Comment:
i suggest the following API for alpha1 so that at minimum the user has the flexibility to set the image from airgap registries.

```
// Size is the expected size of the etcd cluster.
Size int
// Image is the path:tag or path@digest of a container image to use for the etcd cluster members. 
Image string
```

this area of the API later can be extended to the following:
```
type Member struct {
  Name string (omitifempty, should always be prefixed with the cluster name, generated to member1,2,3 if not specififed)
  Image string (omitifempty, overrides defaults)
}

type EtcdClusterSpec struct {
  Name string (name of the cluster)
  DefaultImage string
  Members []Members
}
```

Reply:
> i suggest the following API for alpha1 so that at minimum the user has the flexibility to set the image from airgap registries.
> 
> ```
> // Size is the expected size of the etcd cluster.
> Size int
> // Image is the path:tag or path@digest of a container image to use for the etcd cluster members. 
> Image string
> ```

I understand the motivation of supporting airgap use case.

My original design doc had one more field "registry" to support such case. But removed it to keep v0.1.0 as simple as possible. It's open to add it back or follow your proposal in next step. 

There is also a concern in your proposal, using a digest might be difficult for the operator to compare the version (old version vs new version) when upgrading. We don't support upgrade the etcd cluster across two minor versions, also do not support downgrading for now. So we need to compare the version.



> this area of the API later can be extended to the following:
> 
> ```
> type Member struct {
>   Name string (omitifempty, should always be prefixed with the cluster name, generated to member1,2,3 if not specififed)
>   Image string (omitifempty, overrides defaults)
> }
> 
> type EtcdClusterSpec struct {
>   Name string (name of the cluster)
>   DefaultImage string
>   Members []Members
> }
> ```

I am not sure whether I fully understood your point. Each time when users scale in/out the cluster, users need to update the member list themselves? It seems that it has more flexibility but also a little more complicated. The current design is that users only need to specify a size. 

It's open to discussion for next release, i.e. 0.2.0+.

Could you please raise issues to track your proposals? Thanks

Reply:
> There is also a concern in your proposal, using a digest might be difficult for the operator to compare the version (old version vs new version) when upgrading. We don't support upgrade the etcd cluster across two minor versions, also do not support downgrading for now. So we need to compare the version.

the target cluster and it's members could have a new digest.
that would allow upgrading from one digest -> another.
digest is considered more secure that tag but that argument is not a good one.
so supporting only tags is fine in my book.

> I am not sure whether I fully understood your point. Each time when users scale in/out the cluster, users need to update the member list themselves? It seems that it has more flexibility but also a little more complicated. The current design is that users only need to specify a size.

it's just an idea - to have a current cluster state and a target cluster state.
the target cluster can have a different amount of members e.g. scale to 5, but also making it possible to override the image per member if this operator can be used to verify etcd upgrade or skew.

> Could you please raise issues to track your proposals? Thanks

i think before raising issues we should spend the zoom calls and do API planning / review.

Reply:
> i think before raising issues we should spend the zoom calls and do API planning / review.

Sounds good.

Reply:
> do API planning / review.

@neolit123 would you be able to help draft an API design/plan when you have time?

Reply:
sure, i will draft something for the alpha.